src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+                mMateReferenceIndex = null;
             }
         }
-        // Drop through from above if nothing done.
-        this.mMateReferenceName = mateReferenceName.intern();
-        mMateReferenceIndex = null;
+        else {
+            mMateReferenceName = mateReferenceName.intern();
+            mMateReferenceIndex = null;
+        }
     }
 
     /**
-     * @return index of the reference sequence for this read's mate in the sequence dictionary, or -1
-     * if mate has no reference sequence set.
+     * Returns the mate reference index for this record.
+     *
+     * If the mate reference name for this record has previously been resolved against the sequence dictionary, the
+     * corresponding index is returned directly. Otherwise, the record must have a non-null SAMFileHeader that can be
+     * used to resolve the index for the record's current mate reference name. If the record has a header, and the
+     * name does not appear in the header's sequence dictionary, the value NO_ALIGNMENT_REFERENCE_INDEX (-1) will be
+     * returned. If the record does not have a header, an IllegalArgumentException is thrown.
+     *
+     * @return Index in the sequence dictionary of the mate reference sequence. If the read has no mate reference
+     * sequence, or if the mate reference name is not found in the sequence index, NO_ALIGNMENT_REFERENCE_INDEX (-1)
+     * is returned.
+     *
+     * @throws IllegalArgumentException if the mate reference index cannot be resolved because the SAMFileHeader for the
+     * record is null.
      */
     public Integer getMateReferenceIndex() {
-        if (mMateReferenceIndex == null) {
-            if (mMateReferenceName == null) {
-                mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            } else if (NO_ALIGNMENT_REFERENCE_NAME.equals(mMateReferenceName)){
+        if (null == mMateReferenceIndex) {
+            // try to resolve the reference index
+            if (NO_ALIGNMENT_REFERENCE_NAME.equals(mMateReferenceName)) {
                 mMateReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            } else {
+            }
+            else if (null != mHeader) {
                 mMateReferenceIndex = mHeader.getSequenceIndex(mMateReferenceName);
             }
+            else {
+                throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the mate reference index");
+            }
         }
         return mMateReferenceIndex;
     }
 
     /**
-     * @param referenceIndex Must either equal -1 (indicating no reference), or exist in the sequence dictionary
-     * in the header associated with this record.
-     */
-    public void setMateReferenceIndex(final int referenceIndex) {
-        mMateReferenceIndex = referenceIndex;
-        if (mMateReferenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
+     * Updates the mate reference index. The record must have a valid SAMFileHeader, and the mate reference index must appear in
```

Comment:
"The record must have a valid SAMFileHeader, unless the mateReferenceIndex parameter equals NO_ALIGNMENT_REFERENCE_INDEX"